// components/App.jsx
const GameListAjax = require('./GameListAjax');
const LinkOut = require('./LinkOut');
const Profiles = require('./Profiles');
const React = require('react');

const REGEX_VALID_PROFILE_URL = /^(https?:\/\/)?steamcommunity\.com\/id\/[^\/]+\/?$/;

const getProfileNameFromUrl = (url) => {
  if (url.endsWith('/')) {
    return getProfileNameFromUrl(url.slice(0, -1));
  }
  const iName = url.lastIndexOf('/') + 1;
  return url.slice(iName);
}

module.exports = class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      profileName: '',
      profileUrl: '',
      profileUrlValid: false,
      userList: [],
    };
  }
  setProfileUrl(profileUrl) {
    const profileUrlValid = profileUrl.match(REGEX_VALID_PROFILE_URL);
    const profileName = profileUrlValid ? getProfileNameFromUrl(profileUrl) : '';
    this.setState({
      profileName,
      profileUrl,
      profileUrlValid,
      userList: [],
    });
  }
  onUserListChange(userList) {
    console.log('user list changed:', userList);
    this.setState({userList});
  }
  render() {
    return (
      <main>
        <section className='plain'><h1>What can we play together?</h1></section>
        <section>
          <div><label htmlFor='profile-url'>Steam profile URL</label></div>
          <div>
            <input
              className='profile-url'
              id='profile-url'
              onChange={(e) => this.setProfileUrl(e.target.value)}
              placeholder='steamcommunity.com/id/gaben'
              type='text'
              value={this.state.profileUrl}
            />
          </div>
          <div>
            <label className='help' htmlFor='profile-url'>
              Visit your Steam
              profile <LinkOut url='https://steamcommunity.com/my/profile'>
              here</LinkOut>. When the page loads, copy the URL from the address bar!
            </label>
          </div>
        </section>
        {
          this.state.profileUrl && (
            this.state.profileUrlValid ?
            <Profiles
              name={this.state.profileName}
              onListChange={(list) => this.onUserListChange(list)}
            /> :
            <section>Invalid profile URL.</section>
          )
        }
        {
          this.state.userList.length ? (
            <section>
              <GameListAjax userIdString={this.state.userList.join(',')} />
            </section>
          ) : null
        }
      </main>
    );
  }
};
