// components/GameListAjax.jsx
const api = require('client/api');
const GameList = require('./GameList');
const React = require('react');

module.exports = class GameListAjax extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      games: [],
      loading: true,
    };
  }
  async fetchData(userIdString) {
    this.setState({
      games: [],
      loading: true,
    });
    const games = await api({url: `/games/${userIdString}`});
    // Discard old/invalid requests.
    if (userIdString !== this.props.userIdString) {
      return;
    }
    this.setState({
      games,
      loading: false,
    });
  }
  componentDidMount() {
    this.fetchData(this.props.userIdString);
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.userIdString === this.props.userIdString) {
      return;
    }
    this.fetchData(nextProps.userIdString);
  }
  render() {
    if (this.state.loading) {
      return (
        <div>Loading game list...</div>
      );
    } else {
      return (
        <GameList games={this.state.games} />
      );
    }
  }
}
