// components/Profiles.jsx
const api = require('client/api');
const FriendProfile = require('./FriendProfile');
const { Link } = require('react-router-dom');
const React = require('react');
const UserProfile = require('./UserProfile');
const sharedUtils = require('shared/utils');

module.exports = class Profiles extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errorMessage: null,
      loading: true,
      selected: Object.create(null),
      user: null,
    };
  }
  async fetchData(name) {
    let user = null;
    let errorMessage = null;
    try {
      user = await api({url: `/user/${name}`});
    } catch (e) {
      errorMessage = 'An error occurred. Double-check your profile URL.'
    }
    // Discard old/invalid requests.
    if (name !== this.props.name) {
      return;
    }
    this.setState({
      errorMessage,
      loading: false,
      selected: Object.create(null),
      user,
    });
    this.props.onListChange(user ? [user.id] : []);
  }
  toggleFriend(id) {
    const selected = Object.create(null);
    Object.assign(selected, this.props.selected);
    if (selected[id]) {
      delete selected[id];
    } else {
      selected[id] = true;
    }
    this.setState({selected});
    const list = Object.keys(selected);
    list.push(this.state.user.id);
    list.sort();
    this.props.onListChange(list);
  }
  componentDidMount() {
    this.fetchData(this.props.name);
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.name === this.props.name) {
      return;
    }
    this.setState({
      errorMessage: null,
      loading: true,
      user: null,
    });
    this.fetchData(nextProps.name);
  }
  computeFriendUrl(id) {
    const { pathname } = this.props.location;
    const newSelected = Object.assign(Object.create(null), this.props.selected);
    if (newSelected[id]) {
      // If selected, new URL won't have it selected
      delete newSelected[id];
    } else {
      // If not selected, new URL will have it selected
      newSelected[id] = true;
    }
    const selectedIds = Object.keys(newSelected);
    const search = selectedIds.length ?
      '?selected=' + selectedIds.join(',') :
      '';
    return pathname + search;
  }
  renderInner() {
    if (this.state.loading) {
      return (
        <section>Loading...</section>
      );
    } else if (this.state.user) {
      const sortedFriends = 
        this.state.user.friends.slice().sort((a, b) => sharedUtils.strcmpCaseInsensitive(a.name, b.name));
      return (
        <div>
          <section>
            <UserProfile user={this.state.user} />
          </section>
          <section className='friends'>
            <p>{"Select the friends you'd like to play with."}</p>
            <p className='scroll-down'>{"Scroll down to view the games you have in common."}</p>
            <p><Link to={this.props.location.pathname}>{'Clear selection'}</Link></p>
            <div className='container'>{
             sortedFriends.map((friend) => (
                <FriendProfile
                  friend={friend}
                  key={friend.id}
                  url={this.computeFriendUrl(friend.id)}
                  selected={this.props.selected[friend.id]}
                />
              ))
            }</div>
          </section>
        </div>
      );
    } else {
      return (
        <section>{
          this.state.errorMessage || 'An unknown error occurred.'
        }</section>
      );
    }
  }
  render() {
    return (
      <div className='component-profiles'>{this.renderInner()}</div>
    );
  }
}
